firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,131 @@
+      # write  
+      THEN("the crc accessor method returns 0 before the first write method is called")
+      THEN("The first write method call returns ok status")
+      THEN("After the first write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
+      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
+      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
+      THEN("The output buffer is as expected")
+      # parse
+      THEN("The parse method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload given in the CRCElement constructor is independent of the input body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
+      # write  
+      THEN("The second write method call returns ok status")
+      THEN("After the second write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
+      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
+      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
+      THEN("The output buffer is as expected")
+
+/* CRCElementReceiver
+Scenario: CRCElementReceiver correctly parses CRCElement bodies and performs consistency checking on them
+
+  GIVEN("A CRC element receiver of capacity 254 bytes")
+    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is given to the receiver")
+      THEN("the transform method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload buffer updated in the CRCElement is independent of the input body")
+
+    WHEN("A body without a complete 4-byte header and empty payload is given to the crc element receiver")
+      THEN("the transform status is equal to invalid parse")
+      THEN("the value returned by the crc accessor method is equal to 0x00000000r")
+      THEN("The body's payload section is empty")
+
+    WHEN("body with less than 4 bytes is given to the crc element receiver")
+      // body of length 1,2,3 bytes is passed, passing 4 bytes will give a invalid_crc error instead
+      THEN("the transform status is equal to invalid parse") // for each time
+      THEN("the value returned by the crc accessor method is equal to 0x00000000") // for each time
```

Comment:
I thought about this a bit more and this is not strictly correct (also applies to the same THEN in the previous WHEN). Actually the values returned by the crc and payload methods should just be whatever they were before. For example, if we passed in an output_crcelement with a nonzero crc and a payload with some bytes inside, those should be preserved after transform returns invalid_parse. We should cover this in a test, and I think the WHEN will need to specify what (if anything) was the state of output_crcelement before calling transform.

This test should also specify that the crc accessor method is called on the output_crcelement, not on the CRCElementReceiver object.
This test should also have a THEN checking the value returned by the payload accessor method of output_Crcelement.

Reply:
Correct! so to give some context, for all the tests in CRCElementReceiver I'm using a `output_crcelement` with empty payload given in the constructor, I did that mainly as we are using it as an output parameter and if parse is succesful it will get updated.

> This test should also specify that the crc accessor method is called on the output_crcelement,

will update this for all cases.

So I was thinking for these I'll divide into two GIVEN's , the ones that are already there with
  GIVEN("A CRC element receiver of capacity 254 bytes and output_crcelement constructed with empty payload")
and
  GIVEN("A CRC element receiver of capacity 254 bytes and output_crcelement constructed with payload '0x12 0x13 0x14 0x15 0x16'")

> . For example, if we passed in an output_crcelement with a nonzero crc and a payload with some bytes inside,

Also even with a output_crcelement with some bytes inside, `crc_ ` member parameter is only updated when `write` method is called, so if we want a test that has crcelement with some crc then we must do that.

Reply:
Yes, I agree with the approach of into two GIVENs - I would just specify that the output_crcelement is constructed with an empty payload _buffer_, since the way we're using it in this test is as a place to dump data rather than as the input data itself. In both GIVENs, the THEN could just say that output_crcelement's payload buffer remains unchanged, and the only difference is in the test code assertions.

For the GIVEN with a nonzero CRC, yes part of the test setup code would involve calling `output_crcelement.write` and throwing away the output buffer, in order to make the `crc_` private member become nonzero.

Reply:
If we can establish the following in the CRCElement tests:

THEN the bytes in the constructor's payload buffer look like "..."
THEN the payload accessor method returns the constructor's payload buffer
==
THEN the CRCElement contains the payload "..."

then in the CRCElementReceiver tests we can just assert THEN the CRCElement contains the payload "..."

Reply:
Pattern to use for all CRCElementReceiver tests (and CRCElement tests in a similar pattern):

```
GIVEN ...
  WHEN ...
    THEN the transform method returns ... status code
    THEN the ParsedCRCElement output parameter has the following properties:
      - Its crc accessor method returns ...
      - The payload given to its constructor is ...
      - Its payload accessor method returns a reference to the payload given in its constructor
      - The data in its payload is independent of the data in input_buffer
```